### PR TITLE
Add new card server

### PR DESCRIPTION
**Pull request by nikitasakau**
Decided to use `api/v1/cards` url for this request.

Used:
- `jasmine` for tests
- `request` for post request in tests

### CODE REVIEW

**Comment by nikitasakov** [comment]
[I as user want to add new TODO card](https://trello.com/c/QjJ7VKao/17-i-as-user-want-to-add-new-todo-card)

**Comment by DDKatch** [comment]
LGTM, but @datarockets/cuties what do you think about stubbing database calls in tests, or it's better to setup some db cleaners?

File: server/src/db/migrations/20181129161634-create-card.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+module.exports = {
```
Comment (AleksSenkou) [inline server/src/db/migrations/20181129161634-create-card.js]: File name should be `create_cards` not `create_card`, table names are plural 😉

File: server/src/api/v1/cards.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+const cardsController = require('../../controllers/v1').cards
+
+module.exports = (api) => {
+  api.post('/cards', cardsController.create)
+}
```
Comment (AleksSenkou) [inline server/src/api/v1/cards.js]: @datarockets/cuties the initial thought was that `src/api` is a folder for controllers. Sorry for the misunderstanding, @nikitasakov can we rename it to `src/controllers` in this PR? cause it something that should be here

File: server/src/models/card.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+module.exports = (sequelize, DataTypes) => {
+  const Card = sequelize.define('Card', {
+    description: DataTypes.STRING,
+  }, {})
```
Comment (AleksSenkou) [inline server/src/models/card.js]: @nikitasakov will `id` be acceptable for a Card class?

I mean can I call id of an object or just description?

File: package.json [hunk 4]
```diff
@@ -12,7 +12,8 @@
     "lint": "yarn lint-server && yarn lint-client",
     "lint-server": "cd server && yarn lint",
     "lint-client": "cd client && yarn lint",
-    "update": "bin/update"
+    "update": "bin/update",
+    "test-server": "cd server && yarn jasmine"
```
Comment (AleksSenkou) [inline package.json]: Can we do `test` everywhere for consistency?

What I mean:
`"test-server": "cd server && yarn test"`

and in server/packange.json:

`"test": "yarn jasmine"`

Thoughts?

File: server/spec/cards.spec.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+describe("Post request /api/v1/cards", () => {
+  const request = require("request")
```
Comment (AleksSenkou) [inline server/spec/cards.spec.js]: @nikitasakov can't we do `import request from 'request'` at the top of the file?

File: server/spec/cards.spec.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+      url: base_url,
+      method: "POST",
+      json: true,
+      body: { description: "Test"},
+    }, (error, response, body) => {
+      expect(response.statusCode).toEqual(201)
+      done()
+    });
+  });
+});
```
Comment (AleksSenkou) [inline server/spec/cards.spec.js]: @nikitasakov thoughts about adding a test for checking the negative result? 

`400` code with error message

**Review by AleksSenkou — COMMENTED** [review]
@nikitasakov have you checked it with client code for card creation? it it works well?

File: server/src/api/v1/cards.js (hunk 2 above)
Comment (nikitasakov) [reply server/src/api/v1/cards.js]: @AleksSenkou I've already created `src/controllers` folder, and as I understand `src/api` is a folder with routes. I think we can rename `src/api` to `src/routes`. What do you think?

File: server/src/models/card.js (hunk 3 above)
Comment (nikitasakov) [reply server/src/models/card.js]: @AleksSenkou 
Strange thing :) Look.
First of all I don't know how to test our models like in railsc, because of that I used a crutch: send after creating `.then((card) => { res.status(201).send(card.description) })` and look what it return. 

So, if we send `card.description` we get:


But if we send `card.id` we get an empty json:


If we send `card.createdAt` everything is right again:



And now strange thing: if we send `card.id + card.description` we get and `id` and `description`. I can't understand why it can't return alone id :)

File: server/spec/cards.spec.js (hunk 5 above)
Comment (nikitasakov) [reply server/spec/cards.spec.js]: As I understand - no. At least if I write this way, an errors occurs.

Comment (AleksSenkou) [reply server/spec/cards.spec.js]: @nikitasakov when you say that error occurs, please provide it 😉

File: server/.eslintignore [hunk 7]
```diff
@@ -0,0 +1 @@
+migrations
```
Comment (AleksSenkou) [inline server/.eslintignore]: @nikitasakov you have to config your text editor to leave empty line at the end of each file. Let me knot if it's unclear

File: server/spec/cards.spec.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+describe("Post request /api/v1/cards", () => {
```
Comment (AleksSenkou) [inline server/spec/cards.spec.js]: @datarockets/cuties What do you think about tests folders structure?

Kind of:
```
spec/
  support/
    helpers/
  models/
  repositories/
  features/ (controllers)
  services/ 
  system/
```
?

So `spec/cards.spec.js` would be `spec/contollers/cards.spec.js`

File: server/spec/cards.spec.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+      method: "POST",
+      json: true,
+      body: { description: "Test"},
+    }, (error, response, body) => {
+      expect(response.statusCode).toEqual(201)
+      done()
+    });
+  });
+
+  it("returns 400 code when request dont't have description", (done) => {
```
Comment (AleksSenkou) [inline server/spec/cards.spec.js]: `don't`
I guess should be `doesn't`

File: server/.eslintignore (hunk 7 above)
Comment (AleksSenkou) [reply server/.eslintignore]: Such a problem is in several files

File: server/src/routes/v1/cards.js [hunk 10]
```diff
@@ -0,0 +1,5 @@
+const cardsController = require('../../controllers/v1').cards
+
+module.exports = (api) => {
+  api.post('/cards', cardsController.create)
```
Comment (AleksSenkou) [inline server/src/routes/v1/cards.js]: @nikitasakov can't we use `api.post('/', cardsController.create)` instead?

As I understand that's the advantage of express routing

File: server/.eslintignore (hunk 7 above)
Comment (AleksSenkou) [reply server/.eslintignore]: `"ensure_newline_at_eof_on_save": true,`

for sublime ^

File: server/src/models/card.js (hunk 3 above)
Comment (AleksSenkou) [reply server/src/models/card.js]: @nikitasakov which tool have you used? I'd like to try

Comment (nikitasakov) [reply server/src/models/card.js]: @AleksSenkou Postman

File: server/src/routes/v1/cards.js (hunk 10 above)
Comment (nikitasakov) [reply server/src/routes/v1/cards.js]: @AleksSenkou Nope, he does not see this url 😔.

Comment (AleksSenkou) [reply server/src/routes/v1/cards.js]: @nikitasakov ok, let's go with what we have for now

File: server/src/api/v1/cards.js (hunk 2 above)
Comment (AleksSenkou) [reply server/src/api/v1/cards.js]: @nikitasakov good to me 😉

**Comment by DDKatch** [comment]
> Can you please advise how to implement it?

@nikitasakov @AleksSenkou as I remember we can mock database requests for unit tests in order to keep them fast, but we need to use db for system tests so I found nodejs packages similar with ruby gems:  [factory-girl](https://github.com/aexmachina/factory-girl), [database-cleaner](https://github.com/emerleite/node-database-cleaner). But again they're for system tests, and it's unnecessary if we focus on unit tests now

**Comment by nikitasakov** [comment]
@DDKatch I've tried to test client and server together but got error. Looks like server got request with null description.

I looked client and tried to fix it. If I replace 
```
axios.post('http://localhost:5000/api/v1/cards', {
  description: this.description,
})
```
with
```
axios.post('http://localhost:5000/api/v1/cards', {
  description: this.state.description,
})
```
in
`CreateCardFormContainer/index.js`
And the error is gone. Everything ok. I think if I am right you can fix it in your PR to avoid git conflicts.

**Comment by AleksSenkou** [comment]
@nikitasakov yeah, `this.state.description` should be used 😉

File: server/spec/cards.spec.js (hunk 5 above)
Comment (nikitasakov) [reply server/spec/cards.spec.js]: @AleksSenkou

Comment (DDKatch) [reply server/spec/cards.spec.js]: 🤔

File: server/src/models/card.js (hunk 3 above)
Comment (AleksSenkou) [reply server/src/models/card.js]: @nikitasakov can you give an example of a request as a postman screenshot?

Comment (nikitasakov) [reply server/src/models/card.js]: @AleksSenkou

File: server/spec/cards.spec.js (hunk 5 above)
Comment (DDKatch) [reply server/spec/cards.spec.js]: I suppose jasmine knows nothing about babel, so it crashes. Trying to resolve.

File: server/src/models/card.js (hunk 3 above)
Comment (nikitasakov) [reply server/src/models/card.js]: @AleksSenkou Yes, I got this response too, try to write `.then((card) => { res.status(201).send(card.id) })`

and others variations (`card.createdAt` etc)

Comment (AleksSenkou) [reply server/src/models/card.js]: When sending just `id` it says:

`express deprecated res.send(status): Use res.sendStatus(status) instead src/controllers/v1/cards.js:13:23`

That means that express doesn't understand that you want to send an id because response should be an object, not just value

That's why:

```js
.then((card) => { res.status(201).send({id: card.id}) })
```

will return object with `id` value

Comment (AleksSenkou) [reply server/src/models/card.js]: @datarockets/cuties ^ please, pay more attention on what logs say 😉

Comment (nikitasakov) [reply server/src/models/card.js]: @AleksSenkou But what is the difference between `id` and other parameters? As I understand `description` is not object too, but we can send `card.description`.

Comment (AleksSenkou) [reply server/src/models/card.js]: @nikitasakov because `id` is numeric. And `res.send(number)` previously was used for sending http codes

at the moment `send(number)` is deprecated and it raises a warning and returns `{}`

Comment (AleksSenkou) [reply server/src/models/card.js]: Cool, conversation can be resolved!

File: server/spec/cards.spec.js (hunk 5 above)
Comment (AleksSenkou) [reply server/spec/cards.spec.js]: @datarockets/cuties thoughts about merging with it for now and create an infra task?

@nikitasakov as a PR-holder, please, adjust PR merging as soon as possible so changes can be used for future development 😉

**Review by AleksSenkou — APPROVED** [review]
LGTM!

Good work was done, a lot of new stuff was learnt 🎖

File: server/spec/cards.spec.js (hunk 5 above)
Comment (DDKatch) [reply server/spec/cards.spec.js]: @datarockets/cuties looks like it's `request` package responsibility. I mean in [readme](https://github.com/request/request) and in [examples](https://github.com/request/request/issues/3070) aka everywhere the `const request = require("request")` line is used.
